FIX IterativeImputer skip iterative part if keep_empty_features is set to True

Anurag-Varma opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Fixes #29375


#### What does this implement/fix? Explain your changes.
```python
            if np.any(np.all(np.isnan(X[:, valid_mask]), axis=0)):
                Xt = X
                mask_missing_values[:, valid_mask] = True
            else:
                mask_missing_values[:, valid_mask] = np.logical_or(mask_missing_values[:, valid_mask], np.all(np.isnan(X[:, valid_mask]), axis=0))
                Xt = X
```

Fixes the issues caused when keep_empty_features is set to 'True' and giving different output when compared to that of when set to 'False'. 

#### Any other comments?
After this change, it returns the same result for keep_empty_features set to 'True' or 'False'.

Added 1 new test case to cover the newly added code in if-else condition, by giving the input mentioned in the issue and checking if it's working or not.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
https://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

Anurag-Varma committed:
fixed IterativeImputer - keep_empty_features  bug

Anurag-Varma committed:
linting issues fixed

Anurag-Varma committed:
Added new test cases to cover the newly added code by using different input

Anurag-Varma committed:
fixed linting for newly added test case

Anurag-Varma commented:
Can any member help on how to fix the Changelog ?

I am new to the repo and this is my 2nd contribution

adrinjalali commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -1531,6 +1531,18 @@ def test_iterative_imputer_keep_empty_features(initial_strategy):
+
+    imputer = IterativeImputer(keep_empty_features=keep_empty_features)
+    X_imputed = imputer.fit_transform(X)
+    assert X_imputed.shape == (3, 3)
```
I think a stronger test is to run both True and False in the same run and make sure the results are the same.

Anurag-Varma replied:
Changed the test case to compare the outputs of when keep_empty_features is set to True and False are same or not.

adrinjalali reviewed (commented):
The changelog should go in the `v1.6.rst` file :)

Anurag-Varma committed:
Modified the test case

Modified the test case to check the results are same when keep_empty_features is True and False

Anurag-Varma committed:
Updated changelog

Anurag-Varma commented:
@adrinjalali The changes are done.

If it looks good, can you merge the PR.

adrinjalali commented on doc/about.rst:
the changes on this file are unrelated to the PR.

Anurag-Varma replied:
Sorry, I think this file was added by mistake when merging with the main.

Deleting the file.

adrinjalali commented on doc/whats_new/v1.6.rst:
```diff
@@ -147,6 +147,11 @@ Changelog
   :pr:`29135` by :user:`Xuefeng Xu <xuefeng-xu>`.
 
+- |Fix| :class:`impute.IterativeImputer` Now gives same output 
+  when keep_empty_features is set to True, This is the same as when
+  keep_empty_features is to False.
```
```suggestion
  when ``keep_empty_features=True`` compared to when
  ``keep_empty_features=False`` on non-empty features.
```

adrinjalali reviewed (commented):
Otherwise LGTM. But this would need a second reviewer.

Anurag-Varma committed:
Update doc/whats_new/v1.6.rst

Co-authored-by: Adrin Jalali <adrin.jalali@gmail.com>

xuefeng-xu commented:
@Anurag-Varma Could you test this data below? I add the first column which contains all nan, the rest are the same. I still saw it skip iterative part when set `keep_empty_features=True`.

```python
X = [[np.nan, np.nan, 0, 1], [np.nan, 2, np.nan, 3], [np.nan, 4, 5, np.nan]]
```

Anurag-Varma commented:
> ```python
> X = [[np.nan, np.nan, 0, 1], [np.nan, 2, np.nan, 3], [np.nan, 4, 5, np.nan]]
> ```
Thanks @xuefeng-xu 

Tried this, its failing.
Changed PR to draft, will fix it then update the PR
